### PR TITLE
usersession/client: provide a way for client to send messages to a subset of users

### PR DESCRIPTION
**Pull request by stolowski**
Provide a way for usersession client to send messages only to a subset of clients (based on user IDs).

### CODE REVIEW

**Review by mardy — APPROVED** [review]
LGTM, though I doubt we really need an array of UIDs (one should be enough). On the other hand, handling this as an array doesn't complicate the code, so it's fine.

**Review by MiguelPires — APPROVED** [review]
LGTM

**Comment by stolowski** [comment]
> LGTM, though I doubt we really need an array of UIDs (one should be enough). On the other hand, handling this as an array doesn't complicate the code, so it's fine.

Yeah there is no clear use case for it right now, but at the same time as you said it doesn't complicate the code.

File: usersession/client/client_test.go [hunk 1]
```diff
@@ -448,12 +448,30 @@ func (s *clientSuite) TestServicesStopFailure(c *C) {
 }
 
 func (s *clientSuite) TestPendingRefreshNotification(c *C) {
+	n := 0
 	s.handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		n++
```
Comment (jhenstridge) [inline usersession/client/client_test.go:453]: This increment looks like it is racy. The client will be hitting both session agents in parallel, so if we're unlucky they'll both try to increment `n` at the same time. Maybe using a `sync.Mutex` or `atomic.AddInt32`?

**Review by jhenstridge — CHANGES_REQUESTED** [review]
This looks good. I did notice one small issue in the test that would be worth fixing, but other than that I'm happy with it.

File: usersession/client/client_test.go (hunk 1 above)
Comment (stolowski) [reply usersession/client/client_test.go:453]: Nice catch, thanks, will do.

Comment (stolowski) [reply usersession/client/client_test.go]: Done.

**Comment by stolowski** [comment]
@jhenstridge the test raciness has been fixed, can you take another look?

**Review by jhenstridge — APPROVED** [review]
Looks good!

**Pull request merged**